src/lib.rs

```diff
@@ -1,101 +1,405 @@
+    tree_sitter_grep::Args::parse_from([
         "tree_sitter_grep",
         "-q",
         &aggregated_queries.query_text,
         "-l",
         "rust",
         "--capture",
         CAPTURE_NAME_FOR_TREE_SITTER_GREP,
-    ]);
-    let reported_any_violations = AtomicBool::new(false);
-    tree_sitter_grep::run_with_callback(
-        tree_sitter_grep_args,
-        |capture_info, file_contents, path| {
-            let (rule_index, rule_listener_index) =
-                aggregated_queries.pattern_index_lookup[capture_info.pattern_index];
-            let rule = &resolved_rules[rule_index];
-            let listener = &rule.listeners[rule_listener_index];
-            (listener.on_query_match)(
-                capture_info.node,
-                &QueryMatchContext::new(
-                    path,
-                    file_contents,
-                    rule,
-                    &reported_any_violations,
-                    &context,
-                ),
-            );
+    ])
+}
+
+fn write_files<'a>(files_to_write: impl Iterator<Item = (&'a Path, &'a [u8])>) {
+    for (path, file_contents) in files_to_write {
+        fs::write(path, file_contents).unwrap();
+    }
+}
+
+type RuleName = String;
+
+fn apply_fixes(file_contents: &mut Vec<u8>, pending_fixes: HashMap<RuleName, Vec<PendingFix>>) {
+    let non_conflicting_sorted_pending_fixes =
+        get_sorted_non_conflicting_pending_fixes(pending_fixes);
+    for PendingFix { range, replacement } in non_conflicting_sorted_pending_fixes.into_iter().rev()
+    {
+        file_contents.splice(range, replacement.into_bytes());
+    }
+}
+
+fn compare_pending_fixes(a: &PendingFix, b: &PendingFix) -> Ordering {
+    if a.range.start < b.range.start {
+        return Ordering::Less;
+    }
+    if a.range.start > b.range.start {
+        return Ordering::Greater;
+    }
+    if a.range.end < b.range.end {
+        return Ordering::Less;
+    }
+    if a.range.end > b.range.end {
+        return Ordering::Greater;
+    }
+    Ordering::Equal
+}
+
+fn has_overlapping_ranges(sorted_pending_fixes: &[PendingFix]) -> bool {
+    let mut prev_end = None;
+    for pending_fix in sorted_pending_fixes {
+        if let Some(prev_end) = prev_end {
+            if pending_fix.range.start < prev_end {
+                return true;
+            }
+        }
+        prev_end = Some(pending_fix.range.end);
+    }
+    false
+}
+
+fn get_sorted_non_conflicting_pending_fixes(
+    pending_fixes: HashMap<RuleName, Vec<PendingFix>>,
+) -> Vec<PendingFix> {
+    pending_fixes.into_iter().fold(
+        Default::default(),
+        |accumulated_fixes, (rule_name, mut pending_fixes_for_rule)| {
+            pending_fixes_for_rule.sort_by(compare_pending_fixes);
+            if has_overlapping_ranges(&pending_fixes_for_rule) {
+                panic!("Rule {:?} tried to apply self-conflicting fixes", rule_name);
+            }
+            let mut tentative = accumulated_fixes.clone();
```

Comment:
This is probably not allocation-minimal but trying not to prematurely optimize that